src/nbl/asset/utils/CIESProfile.cpp

```diff
@@ -0,0 +1,300 @@
+
+            *state.fillValue.asUShort = encodeV;
+            state.fillValue.writeMemory(wInfo, blockArrayOffset);
+        };
+
+        CBasicImageFilterCommon::clip_region_functor_t clip(state.subresource, state.outRange, creationParams.format);
+        const auto& regions = outImg->getRegions(state.subresource.mipLevel);
+        CBasicImageFilterCommon::executePerRegion(std::forward<ExecutionPolicy>(policy), outImg.get(), fill, regions.begin(), regions.end(), clip);
+    
+        if constexpr (policy._Parallelize)
+        {
+            integral = aIntegral;
+            nonZeroEmissionDomainSize = aNonZeroEmissionDomainSize;
+        }
+
+       avgEmmision = integral / static_cast<decltype(integral)>(nonZeroEmissionDomainSize);
+    }
+
+    ICPUImageView::SCreationParams viewParams = {};
+    viewParams.image = outImg;
+    viewParams.flags = static_cast<ICPUImageView::E_CREATE_FLAGS>(0);
+    viewParams.viewType = IImageView<ICPUImage>::ET_2D;
+    viewParams.format = viewParams.image->getCreationParameters().format;
+    viewParams.subresourceRange.aspectMask = static_cast<IImage::E_ASPECT_FLAGS>(0);
+    viewParams.subresourceRange.levelCount = viewParams.image->getCreationParameters().mipLevels;
+    viewParams.subresourceRange.layerCount = 1u;
+
+    return ICPUImageView::create(std::move(viewParams));
+}
+
+//! Explicit instantiations
+template core::smart_refctd_ptr<asset::ICPUImageView> CIESProfile::createIESTexture(const std::execution::sequenced_policy&, const size_t&, const size_t&) const;
+// template core::smart_refctd_ptr<asset::ICPUImageView> CIESProfile::createIESTexture(const std::execution::parallel_policy&, const size_t&, const size_t&) const;
+
+core::smart_refctd_ptr<asset::ICPUImageView> CIESProfile::createIESTexture(const size_t& width, const size_t& height) const
+{
+    return createIESTexture(std::execution::seq, width, height); 
+}
+
+template<class ExecutionPolicy>
+bool CIESProfile::flattenIESTexture(ExecutionPolicy&& policy, const IES_STORAGE_FORMAT& avgEmmision, core::smart_refctd_ptr<asset::ICPUImageView> inoutIES, const float flatten)
+{
+    if constexpr (policy._Parallelize)
+        static_assert(false, "meh it appears that filters with par policy are broken on this branch, use seq policy till fixed");
+
+    const bool inFlattenDomain = flatten >= 0.0 && flatten < 1.0; // [0, 1) range for blend equation, 1 is invalid
+
+    if (!inFlattenDomain)
+        return false;
+
+    auto outIES = inoutIES->getCreationParameters().image;
+
+    const auto& creationParams = outIES->getCreationParameters();
+    const bool validFormat = creationParams.format == CIESProfile::IES_TEXTURE_STORAGE_FORMAT;
+
+    if (!validFormat)
+        return false;
+
+    if (flatten > 0.0) // skip if 0 because its just copying texels
+    {
+        CFillImageFilter::state_type state;
+        state.outImage = outIES.get();
```

Comment:
you can't modify the IES you've been passed, needs to be a new ICPUImage+View, with the view inserted into the cache with an extra suffix `?flatten=value` on the key

Reply:
sure

Reply:
now we host the utility in the profile which doesn't have access to cache, so we need to assume the input is already the new view inserted into cache with an extra key `?flatten=value` suffix

Reply:
if we really want to make new asset here with the key and insert into cache in the utility we need to move it into mitsuba sources probably

Reply:
Yes this is why I want `flatten` to just be a parameter for the `CIESProfile createIESTexture` method and get rid of `flattenIESTexture` method

so we can do all this cache stuff outside, in the mitsuba loader

Reply:
I decided to merge flatten texture creation into IES one **but** to keep it into profile because IES loader needs this util too, now we have more params to handle cache insertion